File: apps/admin-web/src/api/identityAdmin.ts
import { http } from "@/utils/http";

export type IdentityCatalog = {
  position_keys: string[];
  appointment_keys: string[];
  scope_types: Array<"UNIT" | "SUBTREE">;
  terminal_statuses: Array<"SUSPENDED" | "ENDED" | "REVOKED">;
  writes_enabled: boolean;
  permission_matrix: Array<{
    role_key: string;
    role_name: string;
    permissions: Array<{
      permission_key: string;
      permission_name: string;
      sensitive_level: "INTERNAL" | "SENSITIVE" | "RESTRICTED";
    }>;
  }>;
};

export type IdentityOrgOption = {
  id: string;
  unit_code: string;
  name: string;
  unit_type: string;
  parent_id?: string;
};

export type PositionAssignment = {
  id: number;
  position_key: string;
  valid_from?: string;
  valid_until?: string;
  status: string;
  source_reference?: string;
};

export type ServiceResponsibility = {
  id: number;
  org_unit_id: string;
  org_name: string;
  scope_type: string;
  valid_from?: string;
  valid_until?: string;
  status: string;
  source_reference?: string;
};

export type Employment = {
  id: number;
  institution_id: string;
  institution_name: string;
  status: string;
  started_on?: string;
  ended_on?: string;
  source_reference?: string;
  positions: PositionAssignment[];
  service_responsibilities: ServiceResponsibility[];
};

export type VolunteerAppointment = {
  id: number;
  appointment_key: string;
  org_unit_id: string;
  org_name: string;
  scope_type: string;
  starts_at: string;
  ends_at: string;
  status: string;
  source_reference: string;
};

export type TechnicalAssignment = {
  id: number;
  assignment_purpose: string;
  starts_at: string;
  ends_at: string;
  status: string;
  source_reference: string;
};

export type IdentityAccount = {
  id: number;
  username: string;
  display_name: string;
  is_active: number;
  is_platform_admin: boolean;
  legacy_roles: string[];
  person_id?: string;
  employments: Employment[];
  volunteer_appointments: VolunteerAppointment[];
  technical_assignments: TechnicalAssignment[];
};

export type ManagedAccount = {
  id: number;
  username: string;
  display_name: string;
  is_active: number;
  last_login_at?: string | null;
  created_at: string;
  roles: string[];
};

type Confirmation = {
  source_reference: string;
  confirmation_note: string;
};

export const getIdentityCatalog = () =>
  http.request<{ success: boolean; data: IdentityCatalog }>(
    "get",
    "/api/v1/identity-admin/catalog"
  );
export const getIdentityAccounts = () =>
  http.request<{ success: boolean; data: IdentityAccount[] }>(
    "get",
    "/api/v1/identity-admin/accounts"
  );

export const createIdentityUser = (data: {
  username: string;
  display_name: string;
  password: string;
}) =>
  http.request<{ success: boolean; data: { id: number } }>(
    "post",
    "/api/v1/iam/users",
    { data: { ...data, roles: [], scopes: [] } }
  );

export const getManagedAccounts = () =>
  http.request<{ success: boolean; data: ManagedAccount[] }>(
    "get",
    "/api/v1/iam/users"
  );

export const resetManagedAccountPassword = (
  userId: number,
  data: { password: string; reason: string }
) =>
  http.request<{
    success: boolean;
    data: { id: number; sessions_revoked: boolean };
  }>("post", `/api/v1/iam/users/${userId}/password`, { data });

export const getIdentityOrgOptions = () =>
  http.request<{ success: boolean; data: IdentityOrgOption[] }>(
    "get",
    "/api/v1/identity-admin/org-options"
  );

export const onboardIdentityEmployee = (data: {
  user_id?: number;
  new_account?: {
    username: string;
    display_name: string;
    password: string;
  };
  position_keys: string[];
  started_on: string;
  ended_on: string;
  service_responsibilities: Array<{
    org_unit_id: string;
    scope_type: "UNIT" | "SUBTREE";
  }>;
  source_reference: string;
  confirmation_note: string;
}) =>
  http.request<{
    success: boolean;
    data: {
      user_id: number;
      person_id: string;
      employment_id: number;
      account_created: boolean;
      person_link_created: boolean;
    };
  }>("post", "/api/v1/identity-admin/employees/onboard", { data });

export const initializeAccountIdentity = (userId: number, data: Confirmation) =>
  http.request<{ success: boolean; data: { person_id: string } }>(
    "post",
    `/api/v1/identity-admin/accounts/${userId}/initialize`,
    { data }
  );

export const changeIdentityAccountStatus = (
  userId: number,
  data: { status: "ACTIVE" | "SUSPENDED"; reason: string }
) =>
  http.request<{ success: boolean; data: { id: number; status: string } }>(
    "post",
    `/api/v1/identity-admin/accounts/${userId}/status`,
    { data }
  );

export const createAccountEmployment = (
  userId: number,
  data: Confirmation & {
    position_key?: string;
    position_keys?: string[];
    started_on: string;
    ended_on?: string;
    service_responsibilities: Array<{
      org_unit_id: string;
      scope_type: "UNIT" | "SUBTREE";
    }>;
  }
) =>
  http.request<{ success: boolean; data: { id: number } }>(
    "post",
    `/api/v1/identity-admin/accounts/${userId}/employments`,
    { data }
  );

export const createAccountVolunteerAppointment = (
  userId: number,
  data: Confirmation & {
    appointment_key: string;
    org_unit_id: string;
    scope_type: "UNIT" | "SUBTREE";
    starts_at: string;
    ends_at: string;
  }
) =>
  http.request<{ success: boolean; data: { id: number } }>(
    "post",
    `/api/v1/identity-admin/accounts/${userId}/volunteer-appointments`,
    { data }
  );

export const createAccountTechnicalAssignment = (
  userId: number,
  data: Confirmation & {
    assignment_purpose: string;
    starts_at: string;
    ends_at: string;
  }
) =>
  http.request<{ success: boolean; data: { id: number } }>(
    "post",
    `/api/v1/identity-admin/accounts/${userId}/technical-assignments`,
    { data }
  );

export const changeIdentityAssignmentStatus = (
  assignmentType: string,
  assignmentId: number,
  data: {
    status: "SUSPENDED" | "ENDED" | "REVOKED";
    reason: string;
  }
) =>
  http.request<{ success: boolean; data: { id: number; status: string } }>(
    "post",
    `/api/v1/identity-admin/assignments/${assignmentType}/${assignmentId}/status`,
    { data }
  );

File: apps/platform-api/app/api/identity_admin.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.api.auth import require_permission
from app.services.identity_admin import (
    catalogs,
    change_assignment_status,
    create_employment,
    create_technical_assignment,
    create_volunteer_appointment,
    initialize_person_link,
    list_identity_accounts,
    list_org_options,
)


router = APIRouter(prefix="/api/v1/identity-admin", tags=["identity-admin"])


class ConfirmationPayload(BaseModel):
    source_reference: str = Field(min_length=4, max_length=500)
    confirmation_note: str = Field(min_length=8, max_length=1000)


class ServiceResponsibilityPayload(BaseModel):
    org_unit_id: str = Field(min_length=1, max_length=64)
    scope_type: str = Field(pattern="^(UNIT|SUBTREE)$")


class EmploymentPayload(ConfirmationPayload):
    # ``position_key`` remains for backwards compatibility.  New identity
    # accounts can hold several position templates under one employment.
    position_key: str | None = Field(default=None, min_length=3, max_length=64)
    position_keys: list[str] = Field(default_factory=list, max_length=16)
    started_on: str
    ended_on: str | None = None
    service_responsibilities: list[ServiceResponsibilityPayload] = Field(
        default_factory=list, max_length=32
    )


class VolunteerAppointmentPayload(ConfirmationPayload):
    appointment_key: str = Field(min_length=3, max_length=64)
    org_unit_id: str = Field(min_length=1, max_length=64)
    scope_type: str = Field(pattern="^(UNIT|SUBTREE)$")
    starts_at: str
    ends_at: str


class TechnicalAssignmentPayload(ConfirmationPayload):
    assignment_purpose: str = Field(min_length=6, max_length=500)
    starts_at: str
    ends_at: str


class StatusChangePayload(BaseModel):
    status: str = Field(pattern="^(SUSPENDED|ENDED|REVOKED)$")
    reason: str = Field(min_length=6, max_length=1000)


def _call(function, *args, **kwargs):
    try:
        return function(*args, **kwargs)
    except PermissionError as exc:
        raise HTTPException(403, str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(400, str(exc)) from exc


@router.get("/catalog")
def get_catalog(
    _: dict = Depends(require_permission("iam:manage")),
) -> dict:
    return {"success": True, "data": _call(catalogs)}


@router.get("/accounts")
def get_accounts(
    _: dict = Depends(require_permission("iam:manage")),
) -> dict:
    return {"success": True, "data": _call(list_identity_accounts)}


@router.get("/org-options")
def get_org_options(
    _: dict = Depends(require_permission("iam:manage")),
) -> dict:
    return {"success": True, "data": _call(list_org_options)}


@router.post("/accounts/{user_id}/initialize")
def initialize_account_identity(
    user_id: int,
    payload: ConfirmationPayload,
    actor: dict = Depends(require_permission("iam:manage")),
) -> dict:
    person_id = _call(
        initialize_person_link,
        actor["id"],
        user_id,
        **payload.model_dump(),
    )
    return {"success": True, "data": {"person_id": person_id}}


@router.post("/accounts/{user_id}/employments")
def add_employment(
    user_id: int,
    payload: EmploymentPayload,
    actor: dict = Depends(require_permission("iam:manage")),
) -> dict:
    values = payload.model_dump()
    if payload.position_key:
        values["position_keys"] = [payload.position_key, *payload.position_keys]
    values.pop("position_key", None)
    values["service_responsibilities"] = [
        item.model_dump() for item in payload.service_responsibilities
    ]
    employment_id = _call(
        create_employment,
        actor["id"],
        user_id,
        **values,
    )
    return {"success": True, "data": {"id": employment_id}}


@router.post("/accounts/{user_id}/volunteer-appointments")
def add_volunteer_appointment(
    user_id: int,
    payload: VolunteerAppointmentPayload,
    actor: dict = Depends(require_permission("iam:manage")),
) -> dict:
    appointment_id = _call(
        create_volunteer_appointment,
        actor["id"],
        user_id,
        **payload.model_dump(),
    )
    return {"success": True, "data": {"id": appointment_id}}


@router.post("/accounts/{user_id}/technical-assignments")
def add_technical_assignment(
    user_id: int,
    payload: TechnicalAssignmentPayload,
    actor: dict = Depends(require_permission("iam:manage")),
) -> dict:
    assignment_id = _call(
        create_technical_assignment,
        actor["id"],
        user_id,
        **payload.model_dump(),
    )
    return {"success": True, "data": {"id": assignment_id}}


@router.post("/assignments/{assignment_type}/{assignment_id}/status")
def update_assignment_status(
    assignment_type: str,
    assignment_id: int,
    payload: StatusChangePayload,
    actor: dict = Depends(require_permission("iam:manage")),
) -> dict:
    _call(
        change_assignment_status,
        actor["id"],
        assignment_type=assignment_type,
        assignment_id=assignment_id,
        **payload.model_dump(),
    )
    return {
        "success": True,
        "data": {"id": assignment_id, "status": payload.status},
    }
